File: scripts/sortfasta/sortfasta.cpp
#include <iostream>
#include <vector>
#include <algorithm> // For std::sort
#include <cstdint>
#include <fstream>
#include <unistd.h> // For getopts
#include <sstream> // for std::stringstream

struct Read 
{
	int64_t number;
	std::string sequence;
};

std::vector<std::string> split(const std::string &s)
/* Tokenizes (i.e. isolates words in sentences and adds to vector) similar 
 *  * to .split() function in python */ 
{
        
        std::vector<std::string> elems;
        std::stringstream ss(s);
        std::string item;

        while (std::getline(ss, item, '_')) {
                if (!item.empty()) {
                        elems.push_back(item);
                }
        }

        return elems;
}

bool readsSorter(Read const& lhs, Read const& rhs)	
{
	return lhs.number < rhs.number;
}

std::vector< Read > getReads(std::string inputPath)
{
	std::ifstream input (inputPath, std::ios::in);
	std::string line;
	std::vector< std::string > header;
	std::vector< Read > reads;
	Read read;
	int64_t readNumber;

	while (!input.eof()) {
		readNumber = 0;

		// Read header of sequence
		std::getline(input, line);

		if (line != "") {
			header = split(line);
			readNumber = stoi( header.at(1) );	
		}

		// Read sequence
		std::getline(input, line);	

		if (line != "" && readNumber != 0) {
			read.number = readNumber;
			read.sequence = line;
		}

		// Add read to reads
		reads.push_back(read);
	}

	input.close();

	return reads;
}

void writeSortedReads(std::vector< Read > reads, std::string outputPath)
{
	std::ofstream output (outputPath, std::ios::out);

	for (int index = 0; index < reads.size(); index++) {
		output << ">" << reads.at(index).number << "\n";
		output << reads.at(index).sequence << "\n";
	}

	output.close();
}

int main(int argc, char *argv[])
{
	std::string help = "Sort reads in FASTA file based on read number.\n";
	std::string usage = "Usage: sortfasta [-i input path][-o output path]\n";

	int opt;

	if (argc == 1) {
		std::cerr << usage;
		return 1;
	}

	std::string inputPath = "";
	std::string outputPath = "";

	while ((opt = getopt(argc, argv, "hi:o:")) != -1) {
		switch (opt) {
			case 'h':
				std::cout << help << usage;	
				return 0;
			case 'i':
				// Source maf file name
				inputPath = optarg;
				break;
			case 'o':
				// maf output file name
				outputPath = optarg;
				break;
		}
	}

	bool optsIncomplete = false;

	if (inputPath == "") {
		std::cerr << "Please provide an input path.\n";
		optsIncomplete = true;	
	}
	if (outputPath == "") {
		std::cerr << "Please provide an output path.\n";
		optsIncomplete = true;
	}
	if (optsIncomplete) {
		std::cerr << usage;
		return 1;
	}

	std::cout << "Reading input file...\n";
	std::vector< Read > reads = getReads(inputPath);

	std::cout << "Sorting reads...\n";
	std::sort(reads.begin(), reads.end(), &readsSorter);

	std::cout << "Writing to output file...\n";
	writeSortedReads(reads, outputPath);

	return 0;
}

File: src/preprocessing/sortfasta/sortfasta.py
import sys
import getopt
import re
from operator import itemgetter

def getReads(inputPath):
	'''
	Inputs
	- string inputPath: specifies the path to the input FASTA file
	Returns
	- list of tuples reads: first element in tuple is int read number,
				second element is string DNA sequence
	'''
	with open(inputPath, 'r') as file:
		reads = []
		sequence = ""
		readNum = -1
		header = ""
		for line in file:
			if line[0] == '>':
				if sequence != "":
					reads.append( (readNum, sequence, header) )
				header = line
				readNum = int( re.findall('(\d+)', line)[readNumberIndex] )
				sequence = ""
			else:
				sequence += line.rstrip('\n')
		reads.append( (readNum, sequence, header) )
	return reads

def writeFasta(outputPath, reads):
	'''
	Inputs
	- string outputPath: specifies the write path to the output
			     FASTA file
	- list of tuples reads: first element in tuple is int read number,
				second element is string DNA sequence
	Returns
	- None
	'''
	with open(outputPath, 'w') as file:
		for read in reads:
			header = "%s" % (read[2])
			file.write(header)	
			sequence = "%s\n" % (read[1])
			file.write(sequence)
				 

helpMessage = "Sort FASTA files based on read number."
usageMessage = "[-h help] [-i input FASTA file] [-o output prefix]"

options = "hn:i:o:"

try:
	opts, args = getopt.getopt(sys.argv[1:], options)
except getopt.GetoptError:
	print "Error: unable to read command line arguments."
	sys.exit(2)

if len(sys.argv) == 1:
	print usageMessage
	sys.exit()

inputPath = None
outputPath = None
readNumberIndex = 1

for opt, arg in opts:
	if opt == '-h':
		print helpMessage
		print usageMessage
		sys.exit()
	elif opt == '-i':
		inputPath = arg
	elif opt == '-o':
		outputPath = arg
	elif opt == '-n':
		readNumberIndex = int(arg)

optsIncomplete = False

if inputPath is None:
	print "Please provide the path to the input FASTA file."
	optsIncomplete = True
if outputPath is None:
	print "Please provide the path to the output FASTA file."
	optsIncomplete = True

if optsIncomplete:
	print usageMessage
	sys.exit(2)

reads = getReads(inputPath)
# Sort reads based on read number
reads = sorted(reads, key=itemgetter(0))
writeFasta(outputPath, reads)
